## Pandas 

     Pandas adalah tools analisis dan manipulasi data open source yang cepat, powerful, fleksibel, dan mudah digunakan, pandas dibuat  pemrograman Python.

    Pandas menawarkan struktur data dan banyak operasi untuk memanipulasi data 
    
    Panda memngkinkan banyak operasi manipulasi data seperti penggabungan, pembentukan kembali , 
    pembersihan dan fitur perselisihan data. 
    
    Dalam hubungannya dengan Matplotlib dan Seaborn, Pandas memberikan berbagai peluang untuk analisis visual data tabel.
    
    Library Pandas banyak digunakan untuk ilmu data / analisis data dan tugas ML .
    
    Struktur data utama di Pandas diimplementasikan dengan Series dan DataFrame
    
    DataFrames sangat bagus untuk merepresentasikan data nyata: 
        baris sesuai dengan instance (objek, observasi, dll.), dan kolom sesuai 
        dengan features untuk setiap instance.
    
Series : 
* Series sangat mirip dengan array NumPy (sebenarnya dibangun di atas object array NumPy).
* Apa yang membedakan array NumPy dari series adalah bahwa series dapat memiliki axis label, artinya dapat di indeks oleh label, bukan hanya lokasi nomor. 
* Itu juga tidak perlu menyimpan data numerik, bisa menampung apa saja python object. 



* Series 1-D --Vector 
* DataFrame  2-D  Metric 

In [22]:
pip install pandas

You should consider upgrading via the '/Users/iketutg/MyKetut/workspace_python/__H8-PYTN071/env_pytn071/bin/python -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [23]:
# membuat series kosong
import pandas as pd 
# pd.Series(data, index, dtype, copy)
var = pd.Series()
print(var)

Series([], dtype: float64)


/var/folders/gh/bm95p3ys6sd6jvzbfl061_3c0000gn/T/ipykernel_31670/81939339.py:3: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  var = pd.Series()


In [24]:
# membuat series ndarray tanpa index
import pandas as pd
import numpy as np
datanp = np.array(['x','y','z','a','b'])
dts = pd.Series(datanp)
print(dts)

0    x
1    y
2    z
3    a
4    b
dtype: object


In [25]:
datanp = np.array(['a','b','c','d','e','f'])
dts = pd.Series(datanp,index=[10,11,12,13,14,15])
print(dts)

10    a
11    b
12    c
13    d
14    e
15    f
dtype: object


In [27]:
datadict = {'a':0.7,'b':2.,'c':1.5,'d':10} 
dtsdict = pd.Series(datadict)
print(dtsdict)

a     0.7
b     2.0
c     1.5
d    10.0
dtype: float64


In [29]:
datadict = {'a':0.7,'b':2.,'c':1.5,'d':10} 
dtsdict = pd.Series(datadict,index =['a','b','d','f'])
print(dtsdict)

a     0.7
b     2.0
d    10.0
f     NaN
dtype: float64


In [30]:
dtsscalar = pd.Series(5,index =[0,1,2,3,4,5,6,7,8,9,10,11])
print(dtsscalar)

0     5
1     5
2     5
3     5
4     5
5     5
6     5
7     5
8     5
9     5
10    5
11    5
dtype: int64


In [31]:
datanp = np.array(['a','b','c','d','e','f'])
dts = pd.Series(datanp)
print(dts)

0    a
1    b
2    c
3    d
4    e
5    f
dtype: object


In [34]:
#print(dts[0])
print(dts[0:3]) # 
print(dts[:4])


0    a
1    b
2    c
dtype: object
0    a
1    b
2    c
3    d
dtype: object


In [35]:
datanp = np.array([10,20,30,40,50,60])
dts = pd.Series(datanp,index =['a','b','c','d','e','f'])
print(dts)

a    10
b    20
c    30
d    40
e    50
f    60
dtype: int64


In [38]:
print(dts['d'])
print(dts['d':'f'])

40
d    40
e    50
f    60
dtype: int64


In [40]:
print(dts.values)
dts.index

[10 20 30 40 50 60]


Index(['a', 'b', 'c', 'd', 'e', 'f'], dtype='object')

In [ ]:
# A single label, e.g. 5 or 'a', (note that 5 is interpreted as a *label* of the index, and **never** as an integer position along the index).

# A list or array of labels, e.g. ['a', 'b', 'c'].

# A slice object with labels, e.g. 'a':'f'.

# A boolean array of the same length as the axis being sliced,
#   e.g. [True, False, True].

# An alignable boolean Series. The index of the key will be aligned before
#   masking.

# An alignable Index. The Index of the returned selection will be the input.

# A callable 
dts.loc

In [41]:
dats = pd.Series([1,2,3,4,5,6],index=[1,2,5,6,8,9])
print(dats)
print(dats.index)

1    1
2    2
5    3
6    4
8    5
9    6
dtype: int64
Int64Index([1, 2, 5, 6, 8, 9], dtype='int64')


In [49]:
# print(dats.loc[5])
#print(dats.loc[[2,6,9]])
#print(dats.loc[2:6])
#kondisi1 = dats.values > 4 
#print(dats.loc[kondisi1])
print(dats.iloc[2])
print(dats[2])

3
2


In [ ]:
# Allowed inputs are:

# An integer, e.g. 5.
# A list or array of integers, e.g. [4, 3, 0].
# A slice object with ints, e.g. 1:7.
# A boolean array.
# A callable func
dts.iloc

## Data Frame 

In [57]:
import pandas as pd

datadict = { 'name' : ['Budi','Iwan','Wati','Joko'] ,
              'age' : [30,50,40,35],
              'address':['DKI','Bandung','Bogor','Bekasi'],
              'qualification':['Msc','MA','Mca','Phd']
            }

df = pd.DataFrame(datadict)
print(df)

   name  age  address qualification
0  Budi   30      DKI           Msc
1  Iwan   50  Bandung            MA
2  Wati   40    Bogor           Mca
3  Joko   35   Bekasi           Phd


In [54]:
# df[['name','qualification']]
df.loc[df['qualification'] =='Phd']
 #df.loc[]

name  age address qualification
3  Joko   35  Bekasi           Phd

In [58]:
for i,j in df.iterrows():
    print(i,j)
    print()

0 name             Budi
age                30
address           DKI
qualification     Msc
Name: 0, dtype: object

1 name                Iwan
age                   50
address          Bandung
qualification         MA
Name: 1, dtype: object

2 name              Wati
age                 40
address          Bogor
qualification      Mca
Name: 2, dtype: object

3 name               Joko
age                  35
address          Bekasi
qualification       Phd
Name: 3, dtype: object



In [59]:
dfangka = pd.DataFrame({'x':[1,2], 'y':[10,30]})
print(dfangka)

   x   y
0  1  10
1  2  30


In [60]:
def square(x):
    return x * x

dfsquare = dfangka.apply(square)
print(dfsquare)

   x    y
0  1  100
1  4  900


**Installing Pandas**

python -m pip install pandas

atau

conda install pandas

## Praktikan Pandas

    kita akan menggunakan pandas dengan menganalisis hasil NBA yang diberikan oleh FiveThirtyEight 
    dalam file CSV 17MB.

In [ ]:
pwd # 

In [ ]:
ls ../data/ # !dir 

In [ ]:
pip install pandas

In [61]:
download_url = 'https://raw.githubusercontent.com/fivethirtyeight/data/master/nba-elo/nbaallelo.csv'

In [ ]:
pip install requests

In [62]:
import requests 

In [63]:
response = requests.get(download_url)

KeyboardInterrupt: 

In [6]:
print(type(response.status_code))
print(response.status_code)
if 200 >= response.status_code < 400: 
    print("Success")
else: 
    print("Error")

<class 'int'>
200
Success


In [7]:
with open('nba_all_elo.csv', 'wb') as f:
    f.write(response.content)

In [8]:
import numpy as np    
import pandas as pd  # sesuai  konvensi mengimpor Pandas dalam Python 

In [9]:
pd.__version__


'1.3.5'

In [64]:
# dengan read_csv kita bisa membaca file csv baik melalui 
# protokol http atau langsung di folder kita masing2. 
##df = pd.read_csv("../data/nbaallelo_hactiv8.csv") # ../data/nbaalleo_hactiv8.csv ini diluar direktor saat ini 
#df = pd.read_csv('https://raw.githubusercontent.com/ardhiraka/PFDS_sources/master/nbaallelo.csv')
df = pd.read_csv("nba_all_elo.csv") ##ini direktory saat ini

In [65]:
#melihat berapa banyak data nba:
len(df)

126314

In [66]:
df.shape  # 126.314 baris dan 23 kolom 

(126314, 23)

In [67]:
df # menampilkan semua data

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
1               1  194611010TRH   NBA        1     1947  11/1/1946   
2               2  194611020CHS   NBA        0     1947  11/2/1946   
3               2  194611020CHS   NBA        1     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  ...  win_equiv  opp_id  \
0                1            0     TRH    Huskies  ...  40.294830     NYK   
1                1            0     NYK     Knicks  ...  41.705170     TRH   
2                1            0     CHS      Stags  ...  42.012257     NYK   
3                2            0     NYK     Knicks  ...  40.692783     CHS   
4                1            0     DTF    Falcons  ...  38.864048     WSC   
...            ...          ...     ...        ...  ...        ...     ...   
126309         100            1     CLE  Cavaliers  ...  60.309792     GSW   
126310         102            1     GSW   Warriors  ...  68.013329     CLE   
126311         101            1     CLE  Cavaliers  ...  60.010067     GSW   
126312         102            1     CLE  Cavaliers  ...  59.290245     GSW   
126313         103            1     GSW   Warriors  ...  68.519516     CLE   

         opp_fran  opp_pts  opp_elo_i  opp_elo_n  game_location  game_result  \
0          Knicks       68  1300.0000  1306.7233              H            L   
1         Huskies       66  1300.0000  1293.2767              A            W   
2          Knicks       47  1306.7233  1297.0712              H            W   
3           Stags       63  1300.0000  1309.6521              A            L   
4        Capitols       50  1300.0000  1320.3811              H            L   
...           ...      ...        ...        ...            ...          ...   
126309   Warriors      103  1790.9591  1809.9791              H            L   
126310  Cavaliers       91  1704.3949  1700.7391              H            W   
126311   Warriors      104  1809.9791  1813.6349              A            L   
126312   Warriors      105  1813.6349  1822.2881              H            L   
126313  Cavaliers       97  1700.7391  1692.0859              A            W   

        forecast notes  
0       0.640065   NaN  
1       0.359935   NaN  
2       0.631101   NaN  
3       0.368899   NaN  
4       0.640065   NaN  
...          ...   ...  
126309  0.546572   NaN  
126310  0.765565   NaN  
126311  0.234435   NaN  
126312  0.481450   NaN  
126313  0.518550   NaN  

[126314 rows x 23 columns]

In [69]:
df.head(7) # menampikan 5 baris 

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   
5          3  194611020DTF   NBA        1     1947  11/2/1946           1   
6          4  194611020PRO   NBA        1     1947  11/2/1946           1   

   is_playoffs team_id   fran_id  ...  win_equiv  opp_id      opp_fran  \
0            0     TRH   Huskies  ...  40.294830     NYK        Knicks   
1            0     NYK    Knicks  ...  41.705170     TRH       Huskies   
2            0     CHS     Stags  ...  42.012257     NYK        Knicks   
3            0     NYK    Knicks  ...  40.692783     CHS         Stags   
4            0     DTF   Falcons  ...  38.864048     WSC      Capitols   
5            0     WSC  Capitols  ...  43.135952     DTF       Falcons   
6            0     BOS   Celtics  ...  40.459381     PRO  Steamrollers   

   opp_pts  opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0       68  1300.0000  1306.7233              H            L  0.640065   NaN  
1       66  1300.0000  1293.2767              A            W  0.359935   NaN  
2       47  1306.7233  1297.0712              H            W  0.631101   NaN  
3       63  1300.0000  1309.6521              A            L  0.368899   NaN  
4       50  1300.0000  1320.3811              H            L  0.640065   NaN  
5       33  1300.0000  1279.6189              A            W  0.359935   NaN  
6       59  1300.0000  1305.1542              A            L  0.359935   NaN  

[7 rows x 23 columns]

In [70]:
df.tail()

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  ...  win_equiv  opp_id  \
126309         100            1     CLE  Cavaliers  ...  60.309792     GSW   
126310         102            1     GSW   Warriors  ...  68.013329     CLE   
126311         101            1     CLE  Cavaliers  ...  60.010067     GSW   
126312         102            1     CLE  Cavaliers  ...  59.290245     GSW   
126313         103            1     GSW   Warriors  ...  68.519516     CLE   

         opp_fran  opp_pts  opp_elo_i  opp_elo_n  game_location  game_result  \
126309   Warriors      103  1790.9591  1809.9791              H            L   
126310  Cavaliers       91  1704.3949  1700.7391              H            W   
126311   Warriors      104  1809.9791  1813.6349              A            L   
126312   Warriors      105  1813.6349  1822.2881              H            L   
126313  Cavaliers       97  1700.7391  1692.0859              A            W   

        forecast notes  
126309  0.546572   NaN  
126310  0.765565   NaN  
126311  0.234435   NaN  
126312  0.481450   NaN  
126313  0.518550   NaN  

[5 rows x 23 columns]

In [71]:
#mengonfigurasi Pandas untuk menampilkan semua kolom  (23 kolom saat ini)
pd.set_option("display.max.columns", None) ## pandas config akan menampilkan keseluruhan kolom , default 20 

In [72]:
df.head() ## perhatikan sdh tdk ada kolom ... 

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  
0  0.640065   NaN  
1  0.359935   NaN  
2  0.631101   NaN  
3  0.368899   NaN  
4  0.640065   NaN

Perhatikan data elo_ , elon_n , win_equiv menggunakan digit decimal 4 / precision 

In [ ]:
pd.set_option("display.precision", 2) ## Pandas config precision 2 digit decimal 

In [ ]:
df.head()

kita dapat menemukan beberapa posibilitas.head() dan .tail() lebih lanjut dengan latihan kecil. Dapatkah kalian mencetak tiga baris terakhir DataFrame kalian?

In [ ]:
df.tail(3)

In [ ]:
df.head(6)

##### Dalam file data nbaallelo.csv terdapat variabel berikut:


|Header	|Definition|
|:-----|:-------|
|gameorder|	Play order of game in NBA history |
| game_id	| Unique ID for each game |
|lg_id	| Which league the game was played in |
| `_iscopy` | 	Each row of data is tied to a single team for a single game, so `_iscopy` flags if this game_id has already occured for the opposing team in the same matchup
| year_id	| Season id, named based on year in which the season ended |
| date_game	| Game date |
| is_playoffs	| Flag for playoff games |
| team_id	| Three letter code for team name, from Basketball Reference |
| fran_id	| Franchise id. Multiple team_ids can fall under the same fran_id due to name changes or moves. Interactive is grouped by fran_id. 
| pts |	Points scored by team |
| elo_i	| Team elo entering the game |
| elo_n	| Team elo following the game |
| win_equiv	| Equivalent number of wins in a 82-game season for a team of elo_n quality |
| opp_id |	Team id of opponent |
| opp_fran |	Franchise id of opponent |
| opp_pts |	Points scored by opponent / point pertandingan |
| opp_elo_i	 | Opponent elo entering the game, sebagai team lawan (Elo sebelum main ) | 
| opp_elo_n	| Opponent elo following the game, sebagai team lawan (Elo setelah main) |
| game_location	| Home (H), away (A), or neutral (N) |
| game_result |	Win or loss for team in the team_id column |
| forecast |	Elo-based chances of winning for the team in the team_id column, based on elo  ratings and game location
|notes|	Additional informatio |

#### Getting to Know Your Data

Kalian telah mengimpor file CSV dengan pustaka Pandas Python dan melihat konten kumpulan data kalian untuk pertama kalinya. Sejauh ini, kalian hanya melihat ukuran kumpulan data kalian dan beberapa baris pertama dan terakhirnya. Selanjutnya, kalian akan mempelajari cara memeriksa data kalian secara lebih sistematis.

#### Displaying Data Types

Langkah pertama untuk mengetahui data kalian adalah menjelajahi berbagai tipe data yang dikandungnya. Meskipun kalian bisa memasukkan apa pun ke dalam list, kolom DataFrame berisi nilai dari tipe data tertentu. Saat kalian membandingkan struktur data Pandas dan Python, kalian akan melihat bahwa behavior ini membuat Pandas jauh lebih cepat!



Kalian bisa menampilkan semua kolom dan tipe datanya dengan .info():


In [73]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   gameorder      126314 non-null  int64  
 1   game_id        126314 non-null  object 
 2   lg_id          126314 non-null  object 
 3   _iscopy        126314 non-null  int64  
 4   year_id        126314 non-null  int64  
 5   date_game      126314 non-null  object 
 6   seasongame     126314 non-null  int64  
 7   is_playoffs    126314 non-null  int64  
 8   team_id        126314 non-null  object 
 9   fran_id        126314 non-null  object 
 10  pts            126314 non-null  int64  
 11  elo_i          126314 non-null  float64
 12  elo_n          126314 non-null  float64
 13  win_equiv      126314 non-null  float64
 14  opp_id         126314 non-null  object 
 15  opp_fran       126314 non-null  object 
 16  opp_pts        126314 non-null  int64  
 17  opp_elo_i      126314 non-nul

Kalian akan melihat daftar semua kolom dalam kumpulan data kalian dan jenis data yang ada di setiap kolom. Di sini, kalian bisa melihat tipe data int64, float64, dan object. Pandas menggunakan library NumPy untuk bekerja dengan tipe ini. Nanti, kalian akan menemukan tipe data kategorikal yang lebih kompleks, yang diimplementasikan oleh pustaka Pandas.

*Perhatikan tipe data object dimana jika string akan menjadi object .  *Di library numpy tidak ada tipe data string, sehingga menggunakan generik yaitu object .* 


#### Showing Basics Statistics
    Setelah kalian melihat jenis data apa yang ada dalam kumpulan data kalian , 
    sekarang saatnya untuk mendapatkan overview tentang nilai yang terkandung 
    di setiap kolom. Kalian dapat melakukan ini dengan .describe():



In [74]:
df.describe()

gameorder        _iscopy        year_id     seasongame  \
count  126314.000000  126314.000000  126314.000000  126314.000000   
mean    31579.000000       0.500000    1988.200374      43.533733   
std     18231.927643       0.500002      17.582309      25.375178   
min         1.000000       0.000000    1947.000000       1.000000   
25%     15790.000000       0.000000    1975.000000      22.000000   
50%     31579.000000       0.500000    1990.000000      43.000000   
75%     47368.000000       1.000000    2003.000000      65.000000   
max     63157.000000       1.000000    2015.000000     108.000000   

         is_playoffs            pts          elo_i          elo_n  \
count  126314.000000  126314.000000  126314.000000  126314.000000   
mean        0.063857     102.729982    1495.236055    1495.236055   
std         0.244499      14.814845     112.139945     112.461687   
min         0.000000       0.000000    1091.644500    1085.774400   
25%         0.000000      93.000000    1417.237975    1416.994900   
50%         0.000000     103.000000    1500.945550    1500.954400   
75%         0.000000     112.000000    1576.060000    1576.291625   
max         1.000000     186.000000    1853.104500    1853.104500   

           win_equiv        opp_pts      opp_elo_i      opp_elo_n  \
count  126314.000000  126314.000000  126314.000000  126314.000000   
mean       41.707889     102.729982    1495.236055    1495.236055   
std        10.627332      14.814845     112.139945     112.461687   
min        10.152501       0.000000    1091.644500    1085.774400   
25%        34.103035      93.000000    1417.237975    1416.994900   
50%        42.113357     103.000000    1500.945550    1500.954400   
75%        49.635328     112.000000    1576.060000    1576.291625   
max        71.112038     186.000000    1853.104500    1853.104500   

            forecast  
count  126314.000000  
mean        0.500000  
std         0.215252  
min         0.020447  
25%         0.327989  
50%         0.500000  
75%         0.672011  
max         0.979553

-------------

https://en.wikipedia.org/wiki/Quantile 

* count - The number of not-empty values.
* mean - The average (mean) value.
* std - The standard deviation.
* min - the minimum value.
* 25% - The 25% percentile*.
* 50% - The 50% percentile*.
* 75% - The 75% percentile*.
* max - the maximum value.

*Percentile : 25% berapa banyak nilai yang lebih kecil dari persentil yang diberikan.
  numbers between: 0 and 1

In [ ]:
df.quantile(.50)

In [ ]:
s = pd.Series([1, 2, 3, 1])
s.describe()

In [ ]:
print(s.quantile(.25))
print(s.quantile(.50))
print(s.quantile(.75))

In [ ]:
#https://en.wikipedia.org/wiki/Quantile 
sarr = np.sort([1, 2, 3, 1])
print(sarr)
sidx = (sarr.size-1)*.25
print(sidx)
print(sarr[1])
print(sarr[0])
print( sarr[0] + (sarr[1]-sarr[0]))
print( sarr[0] + (sarr[1]-sarr[0]) * .75)
print("Percentile .25% : ",sarr[0] + (sarr[1]-sarr[0]) * .75 )
seria = pd.Series(sarr)
print("Percentile .25% use pandas series quantile func : ",seria.quantile(.25)) 
print("Percentile .25% use numpy quantile func : ",np.percentile(sarr,.25)) 

In [ ]:
## percentile 

arr = np.sort(np.random.randint(33,size=50))
print(arr)
pindex = (arr.size -1) * .37
print(pindex) # 18.13
## no value exis at arr [18.13] there is value at arr[18] and arr[19] 
## linear slop = between 18 and 19 position (arr[19]-arr[18])
# value 18.13 is = value at 18 + linear slop * .13 
print(arr[18] + (arr[19]-arr[18]) * .13)

np.percentile(arr, 37)

------------

#### Exploring Your Dataset
Exploratory data analysis dapat membantu kalian menjawab pertanyaan tentang kumpulan data kalian. Misalnya, kalian dapat memeriksa seberapa sering nilai tertentu muncul dalam kolom:

In [75]:
df

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
1               1  194611010TRH   NBA        1     1947  11/1/1946   
2               2  194611020CHS   NBA        0     1947  11/2/1946   
3               2  194611020CHS   NBA        1     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts      elo_i      elo_n  \
0                1            0     TRH    Huskies   66  1300.0000  1293.2767   
1                1            0     NYK     Knicks   68  1300.0000  1306.7233   
2                1            0     CHS      Stags   63  1300.0000  1309.6521   
3                2            0     NYK     Knicks   47  1306.7233  1297.0712   
4                1            0     DTF    Falcons   33  1300.0000  1279.6189   
...            ...          ...     ...        ...  ...        ...        ...   
126309         100            1     CLE  Cavaliers   82  1723.4149  1704.3949   
126310         102            1     GSW   Warriors  104  1809.9791  1813.6349   
126311         101            1     CLE  Cavaliers   91  1704.3949  1700.7391   
126312         102            1     CLE  Cavaliers   97  1700.7391  1692.0859   
126313         103            1     GSW   Warriors  105  1813.6349  1822.2881   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
0       40.294830    NYK     Knicks       68  1300.0000  1306.7233   
1       41.705170    TRH    Huskies       66  1300.0000  1293.2767   
2       42.012257    NYK     Knicks       47  1306.7233  1297.0712   
3       40.692783    CHS      Stags       63  1300.0000  1309.6521   
4       38.864048    WSC   Capitols       50  1300.0000  1320.3811   
...           ...    ...        ...      ...        ...        ...   
126309  60.309792    GSW   Warriors      103  1790.9591  1809.9791   
126310  68.013329    CLE  Cavaliers       91  1704.3949  1700.7391   
126311  60.010067    GSW   Warriors      104  1809.9791  1813.6349   
126312  59.290245    GSW   Warriors      105  1813.6349  1822.2881   
126313  68.519516    CLE  Cavaliers       97  1700.7391  1692.0859   

       game_location game_result  forecast notes  
0                  H           L  0.640065   NaN  
1                  A           W  0.359935   NaN  
2                  H           W  0.631101   NaN  
3                  A           L  0.368899   NaN  
4                  H           L  0.640065   NaN  
...              ...         ...       ...   ...  
126309             H           L  0.546572   NaN  
126310             H           W  0.765565   NaN  
126311             A           L  0.234435   NaN  
126312             H           L  0.481450   NaN  
126313             A           W  0.518550   NaN  

[126314 rows x 23 columns]

In [76]:
# Metode .value_counts() akan menghitung berapa kali 
# setiap nilai muncul di kolom tertentu, team_id dalam contoh ini.
df["team_id"].value_counts() # 
# team yg sering dan jarang 

BOS    5997
NYK    5769
LAL    5078
DET    4985
PHI    4533
       ... 
INJ      60
PIT      60
DTF      60
TRH      60
SDS      11
Name: team_id, Length: 104, dtype: int64

In [77]:
df["fran_id"].value_counts()

Lakers          6024
Celtics         5997
Knicks          5769
Warriors        5657
Pistons         5650
Sixers          5644
Hawks           5572
Kings           5475
Wizards         4582
Spurs           4309
Bulls           4307
Pacers          4227
Thunder         4178
Rockets         4154
Nuggets         4120
Nets            4106
Suns            4080
Bucks           4034
Trailblazers    3870
Cavaliers       3810
Clippers        3733
Jazz            3555
Mavericks       3013
Heat            2371
Pelicans        2254
Magic           2207
Timberwolves    2131
Grizzlies       1657
Raptors         1634
Hornets          894
Colonels         846
Squires          799
Spirits          777
Stars            756
Sounds           697
Baltimore        467
Floridians       440
Condors          430
Capitols         291
Olympians        282
Sails            274
Stags            260
Bombers          249
Steamrollers     168
Packers           72
Redskins          65
Rebels            63
Waterloo     

Tampaknya tim bernama "Lakers" memainkan .... pertandingan, 

In [78]:
df['fran_id'] == 'Lakers' 
# ini akan menampilkan serangkaian data boolean false / true 
# ini kita anggap sebgai daftar index

0         False
1         False
2         False
3         False
4         False
          ...  
126309    False
126310    False
126311    False
126312    False
126313    False
Name: fran_id, Length: 126314, dtype: bool


|notes    | loc    | 
|:--------|:-------|
|a value   | ```loc['label'] or loc[0]``` <br> single label or integer  |
|a list |  ```loc[['label','label2]] ```  <br>  A list of labe      | 
| slicing |  ```loc[['A':'B']] ```  <br> A and B Include     | 
| Condition |   bool series or list        | 
| Callable function | ```loc[lambda A: A[2]] ```          | 


In [79]:
df.loc[df['fran_id'] == 'Lakers'] # 6024 rows dan menampilkan semua columns 

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
1136          569  194811040BLB   NBA        1     1949   11/4/1948   
1152          577  194811060WSC   NBA        1     1949   11/6/1948   
1159          580  194811090BOS   NBA        1     1949   11/9/1948   
1170          586  194811110NYK   NBA        1     1949  11/11/1948   
1183          592  194811130PRO   NBA        1     1949  11/13/1948   
...           ...           ...   ...      ...      ...         ...   
126016      63009  201504080DEN   NBA        1     2015    4/8/2015   
126052      63027  201504100LAL   NBA        0     2015   4/10/2015   
126086      63044  201504120LAL   NBA        0     2015   4/12/2015   
126115      63058  201504130SAC   NBA        1     2015   4/13/2015   
126137      63069  201504150LAL   NBA        0     2015   4/15/2015   

        seasongame  is_playoffs team_id fran_id  pts      elo_i      elo_n  \
1136             1            0     MNL  Lakers   84  1527.0000  1542.2545   
1152             2            0     MNL  Lakers   62  1542.2545  1536.2078   
1159             3            0     MNL  Lakers   55  1536.2078  1513.4880   
1170             4            0     MNL  Lakers   77  1513.4880  1524.9596   
1183             5            0     MNL  Lakers   90  1524.9596  1533.2726   
...            ...          ...     ...     ...  ...        ...        ...   
126016          78            0     LAL  Lakers  101  1313.7802  1307.5118   
126052          79            0     LAL  Lakers  106  1307.5118  1312.9532   
126086          80            0     LAL  Lakers  106  1312.9532  1305.9197   
126115          81            0     LAL  Lakers   92  1305.9197  1301.3918   
126137          82            0     LAL  Lakers   99  1301.3918  1283.2395   

        win_equiv opp_id      opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
1136    51.218754    BLB     Baltimore       72  1507.2052  1491.9507   
1152    50.621624    WSC      Capitols       67  1481.8721  1487.9188   
1159    48.341133    BOS       Celtics       77  1359.3345  1382.0542   
1170    49.499451    NYK        Knicks       68  1460.4232  1448.9517   
1183    50.330185    PRO  Steamrollers       67  1245.9456  1237.6326   
...           ...    ...           ...      ...        ...        ...   
126016  21.921890    DEN       Nuggets      119  1415.0565  1421.3248   
126052  22.371969    MIN  Timberwolves       98  1283.9249  1278.4835   
126086  21.791288    DAL     Mavericks      120  1556.2162  1563.2498   
126115  21.422607    SAC         Kings      102  1395.6000  1400.1278   
126137  19.985550    SAC         Kings      122  1400.1278  1418.2802   

       game_location game_result  forecast notes  
1136               A           W  0.386583   NaN  
1152               A           L  0.443232   NaN  
1159               A           L  0.608859   NaN  
1170               A           W  0.432863   NaN  
1183               A           W  0.737010   NaN  
...              ...         ...       ...   ...  
126016             A           L  0.238914   NaN  
126052             H           W  0.670716   NaN  
126086             H           L  0.304769   NaN  
126115             A           L  0.251263   NaN  
126137             H           L  0.501819   NaN  

[6024 rows x 23 columns]

In [80]:
df.loc[df['fran_id'] == 'Lakers',"team_id"] # 6024 rows dan hanya menampilkan teamid 

1136      MNL
1152      MNL
1159      MNL
1170      MNL
1183      MNL
         ... 
126016    LAL
126052    LAL
126086    LAL
126115    LAL
126137    LAL
Name: team_id, Length: 6024, dtype: object

In [81]:
df.loc[df["fran_id"] == "Lakers", "team_id"].value_counts() # grouping dan di jumlahkan 

LAL    5078
MNL     946
Name: team_id, dtype: int64

Tampaknya tim bernama "Lakers" memainkan 6024 pertandingan, tetapi hanya 5078 yang dimainkan oleh Los Angeles Lakers. 

Kita akan mencari tahu kenapa demikian :

In [82]:
df.loc[df['team_id']=='MNL', 'date_game'] 

1136     11/4/1948
1152     11/6/1948
1159     11/9/1948
1170    11/11/1948
1183    11/13/1948
           ...    
9701     3/19/1960
9702     3/20/1960
9709     3/22/1960
9711     3/24/1960
9715     3/26/1960
Name: date_game, Length: 946, dtype: object

Memang, Minneapolis Lakers ("MNL") memainkan 946 pertandingan. Kalian bahkan dapat mengetahui kapan mereka memainkan game-game itu:

In [83]:
df.loc[df["team_id"] == "MNL", "date_game"].max() # '4/9/1959' ini adalah output yg salah

'4/9/1959'

Namun, string dan tanggal ditafsirkan secara berbeda oleh Python, dan Pandas tidak secara implisit melemparkan string dalam format tanggal ke tanggal.

Pandas menyediakan fungsi untuk untuk mengubah kolom string menjadi tanggal . Fungsi **pd.to_datetime()** dengan asumsi string adalah format tanggal yang valid.


Agar Anda dapat membandingkan string dan tanggal, simpan tanggal di kolom baru. Membuat kolom baru pada DataFrame semudah menetapkan nama kolom. Sekarang jika Anda menjalankan kode yang sama lagi tetapi pada kolom 'date_played', ia mengembalikan tanggal 26 Maret 1960. Namun, karena representasi string 26 Maret 1960, yang akan menjadi '26/3/1960', mendahului string '4/9/1959', yang terakhir salah dikembalikan sebagai tanggal terbaru di kolom 'date_game'.

In [84]:
df['date_played'] = pd.to_datetime(df['date_game'])
# kita membuat kolom baru dengan tipe date 

In [85]:
df.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes date_played  
0  0.640065   NaN  1946-11-01  
1  0.359935   NaN  1946-11-01  
2  0.631101   NaN  1946-11-02  
3  0.368899   NaN  1946-11-02  
4  0.640065   NaN  1946-11-02

In [86]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 24 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   gameorder      126314 non-null  int64         
 1   game_id        126314 non-null  object        
 2   lg_id          126314 non-null  object        
 3   _iscopy        126314 non-null  int64         
 4   year_id        126314 non-null  int64         
 5   date_game      126314 non-null  object        
 6   seasongame     126314 non-null  int64         
 7   is_playoffs    126314 non-null  int64         
 8   team_id        126314 non-null  object        
 9   fran_id        126314 non-null  object        
 10  pts            126314 non-null  int64         
 11  elo_i          126314 non-null  float64       
 12  elo_n          126314 non-null  float64       
 13  win_equiv      126314 non-null  float64       
 14  opp_id         126314 non-null  object        
 15  

In [87]:
df.loc[df['team_id']=='MNL','date_played'].max()

Timestamp('1960-03-26 00:00:00')

In [ ]:
df.info()

 date_played    126314 non-null  datetime64[ns]

In [88]:
df.loc[df["team_id"] == "MNL", "date_game"].min() ## tidak sesuai

'1/1/1949'

In [89]:
df.loc[df['team_id']=='MNL','date_played'].min()

Timestamp('1948-11-04 00:00:00')

In [ ]:
df.loc[df["team_id"] == "MNL", "date_game"].agg(("min", "max")) ## tidak sesuai

In [90]:
df.loc[df['team_id']=='MNL','date_played'].agg(("min", "max"))

min   1948-11-04
max   1960-03-26
Name: date_played, dtype: datetime64[ns]

------------

Sepertinya Minneapolis Lakers dimainkan main tahun 1948 dan 1960. Itu menjelaskan mengapa kalian mungkin tidak mengenali tim ini!

Kalian juga telah mengetahui mengapa tim Boston Celtics "BOS" memainkan game paling banyak dalam kumpulan data. Mari kita analisis sedikit tentang sejarah mereka. Cari tahu berapa banyak poin yang telah dicetak Boston Celtics selama semua pertandingan yang terdapat dalam dataset ini.

In [91]:
df.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes date_played  
0  0.640065   NaN  1946-11-01  
1  0.359935   NaN  1946-11-01  
2  0.631101   NaN  1946-11-02  
3  0.368899   NaN  1946-11-02  
4  0.640065   NaN  1946-11-02

In [92]:
df.loc[df["team_id"] == "BOS", "pts"].sum()
##Boston Celtics mencetak total 626.484 poin.
## tapi berapa jumlah rata poin per tahun ? 

626484

Boston Celtics mencetak total 626.484 poin.
*tapi berapa jumlah rata poin per tahun ?* 

Pertama, dapatkan tanggal permainan yang dimainkan oleh Boston Celtics, dan Anda sudah tahu cara melakukannya. Anda hanya tertarik pada tahun, sehingga Anda dapat menerapkan fungsi lambda ke setiap tanggal.


Untuk mendapatkan jumlah poin lagi dan menyimpannya, Anda cukup memodifikasi sel yang memanggil metode .sum().

Sekarang Anda dapat menghitung poin rata-rata per tahun.

In [93]:
df.loc[df['team_id']=='BOS','date_played']

6        1946-11-02
14       1946-11-05
28       1946-11-09
40       1946-11-13
46       1946-11-14
            ...    
126141   2015-04-15
126162   2015-04-19
126173   2015-04-21
126185   2015-04-23
126205   2015-04-26
Name: date_played, Length: 5997, dtype: datetime64[ns]

In [94]:
df.loc[df['team_id']=='BOS','date_played'].apply(lambda x: x.year)
## fungsi apply menereruskan setiap nilai ke lambda
## ini terdapat tahun2 yang duplikat 

6         1946
14        1946
28        1946
40        1946
46        1946
          ... 
126141    2015
126162    2015
126173    2015
126185    2015
126205    2015
Name: date_played, Length: 5997, dtype: int64

  #### data frame example apply 

Fungsi apply dimanfaatkan saat kita perlu melakukan membuat/memanipulasi/mengubah/melakukan kalkulasi kolom atau baris pada pandas.

In [95]:
dfc = pd.DataFrame({'A': [1, 2], 'B': [10, 20]})

In [ ]:
def square(x):
    return x * x


In [96]:
dfc

A   B
0  1  10
1  2  20

In [ ]:
df1 = dfc.apply(square)
df1 

In [97]:
df2 = dfc.apply(lambda x: x * 3)
df2

A   B
0  3  30
1  6  60

Fungsi .apply() akan meneruskan setiap nilai ke lambda dan kemudian yang perlu Anda lakukan hanyalah mengekstrak atribut .year.
Terdapat tahun yang duplikat kita hapus dengan panggil metode .unique(),


In [98]:
df.loc[df['team_id']=='BOS','date_played'].apply(lambda v: v.year).unique()
# kita mendapatkan data tahun yang uniq dari boston 

array([1946, 1947, 1948, 1949, 1950, 1951, 1952, 1953, 1954, 1955, 1956,
       1957, 1958, 1959, 1960, 1961, 1962, 1963, 1964, 1965, 1966, 1967,
       1968, 1969, 1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978,
       1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989,
       1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000,
       2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011,
       2012, 2013, 2014, 2015])

In [99]:
 # kita siapkan variabel total point boston
total_poin_BOS = df.loc[df["team_id"] == "BOS", "pts"].sum()
total_poin_BOS

626484

In [102]:
tahun_main_BOS = len(df.loc[df['team_id']=='BOS','date_played'].apply(lambda v: v.year).unique())
print(tahun_main_BOS)
print(type(tahun_main_BOS))

70
<class 'int'>


In [103]:
total_poin_BOS / tahun_main_BOS  # rata rata point tiap tahun 

8949.771428571428


Kalian dapat melihat konsep ini diimplementasikan dengan kumpulan data NBA yang lebih besar. Apakah dataset tersebut berisi kolom yang disebut "points", atau apakah disebut "pts"?

Karena kalian tidak menentukan kolom indeks saat kalian membaca di file CSV, Pandas telah menetapkan RangeIndex ke DataFrame:

In [104]:
df.index # 


RangeIndex(start=0, stop=126314, step=1)

In [105]:
df.axes  # df, seperti semua objek DataFrame, memiliki dua axes:


[RangeIndex(start=0, stop=126314, step=1),
 Index(['gameorder', 'game_id', 'lg_id', '_iscopy', 'year_id', 'date_game',
        'seasongame', 'is_playoffs', 'team_id', 'fran_id', 'pts', 'elo_i',
        'elo_n', 'win_equiv', 'opp_id', 'opp_fran', 'opp_pts', 'opp_elo_i',
        'opp_elo_n', 'game_location', 'game_result', 'forecast', 'notes',
        'date_played'],
       dtype='object')]

In [106]:
df.axes[0]

RangeIndex(start=0, stop=126314, step=1)

Kalian dapat memeriksa keberadaan kolom dengan .keys():

In [107]:
"points" in df.keys()

False

In [108]:
"pts" in df.keys()

True

#### Getting to Know Pandas’ Data Structures
Meskipun DataFrame menyediakan fungsi yang bisa terasa cukup intuitif, konsep dasarnya agak lebih rumit untuk dipahami. Oleh karena itu, kalian akan menyisihkan NBA DataFrame yang luas dan membuat beberapa objek Pandas yang lebih kecil dari awal.

#### Understanding Series Objects
Struktur data Python yang paling dasar adalah list, yang juga merupakan titik awal yang baik untuk mengenal objek pandas.Series. Buat objek Series baru menggunakan list:


The Series mempunyai 2 part : values and identifiers. 
* The values are a sequence, similar to a list in Python. 
* The identifiers are mapped to the values. with index

In [ ]:
revenues = pd.Series([5555, 7000, 1980])

Kalian menggunakan list [5555, 7000, 1980] untuk membuat objek Series yang disebut revenues. Objek Series mempunyai dua komponen:

    * A sequence of values
    * A sequence of identifiers, which is the index
Kalian dapat mengakses komponen ini masing-masing dengan .values dan.index:

In [ ]:
revenues

revenues.values mengembalikan nilai dalam Seri, sedangkan revenues.index mengembalikan indeks posisi.

In [ ]:
revenues.values

In [ ]:
revenues.index

Meskipun Pandas dibuat di atas NumPy, perbedaan yang signifikan terletak pada pengindeksannya. 

Sama seperti array NumPy, Pandas Series juga memiliki indeks integer yang ditentukan secara implisit. 

Indeks implisit ini menunjukkan posisi elemen dalam Series.

Namun, Series juga dapat memiliki jenis indeks yang berubah-ubah. Kalian dapat menganggap indeks eksplisit ini sebagai label untuk baris tertentu:



In [ ]:
city_revenues = pd.Series( [4200, 8000, 6500], index=["Amsterdam", "Toronto", "Tokyo"] )
city_revenues

Di sini, indeks adalah daftar nama kota yang diwakili oleh string. Kalian mungkin telah memperhatikan bahwa dictionaries Python menggunakan indeks string juga, dan ini adalah analogi yang berguna untuk diingat! 

Kalian dapat menggunakan blok kode di atas untuk membedakan antara dua jenis Series:

Setiap series akan menyimpan indeks numerik secara default . ini sama halnya dengan tipe koleksi python primitif List 

revenues: Seri ini berperilaku seperti list Python karena hanya memiliki indeks posisi. 

city_revenues: Seri ini bertindak seperti dictonary Python karena kita dpt menggunakan indeks untuk mengambil value terkait, atau posisi dan label.

Berikut cara membuat Series dengan indeks label dari dictionary Python:



In [ ]:
revenues[0]

In [ ]:
city_revenues['Tokyo']

In [ ]:
city_emp_data = {"Amsterdam": 5, "Tokyo": 8}
city_employee_count = pd.Series(city_emp_data)
city_employee_count

Dictionary keys menjadi indeks, dan nilai dictionary adalah nilai Series.

Sama seperti dictionaries, Series juga mendukung .keys() dan keyword in:

In [ ]:
city_employee_count.keys()

In [ ]:
"Tokyo" in city_employee_count

In [ ]:
"New York" in city_employee_count



#### Understanding DataFrame Objects

Meskipun Series adalah struktur data yang cukup kuat, ia memiliki keterbatasan. Misalnya, kalian hanya dapat menyimpan satu atribut per kunci. Seperti yang kalian lihat dengan kumpulan data nba, yang menampilkan 23 kolom, pustaka Pandas Python memiliki lebih banyak hal untuk ditawarkan dengan DataFrame-nya. Struktur data ini adalah sequence objek kalian yang memiliki indeks yang sama.

Jika kalian telah mengikuti contoh Series diatas, kalian seharusnya sudah memiliki dua objek Series dengan cities sebagai kuncinya:

* city_revenues
* city_employee_count

Kalian dapat menggabungkan objek-objek ini ke dalam DataFrame dengan memberikan dictionary di konstruktor. Dictionary keys akan menjadi nama kolom, dan nilainya akan berisi objek Series:

In [ ]:
city_revenues

In [ ]:
city_employee_count

In [ ]:
city_data = pd.DataFrame({
    "revenue": city_revenues,
    "employee_count": city_employee_count
})
city_data

Perhatikan bagaimana Pandas mengganti nilai employee_count yang hilang untuk Toronto dengan NaN.

Indeks DataFrame baru adalah gabungan dari dua indeks Series:

In [ ]:
city_data.index

Sama seperti Series, DataFrame juga menyimpan nilainya dalam array NumPy:



In [ ]:
city_data.values

Kalian juga dapat merujuk ke 2 dimensi DataFrame sebagai axes:



In [ ]:
city_data.axes?

In [ ]:
city_data.axes

In [ ]:
city_data.axes[0]

In [ ]:
city_data.axes[1]

In [ ]:
pd.DataFrame.axes?

* Axis bertanda 0 adalah indeks baris, 
* Axis bertanda 1 adalah indeks kolom. 

Terminologi ini penting untuk diketahui karena kalian akan menemukan beberapa metode DataFrame yang menerima parameter axis.

DataFrame juga merupakan struktur data seperti dictionary, sehingga juga mendukung .keys() dan keyword in. Namun, untuk DataFrame, hal ini tidak terkait dengan indeks, tetapi dengan kolom:



In [ ]:
'tokyo' in city_data.axes[0]

In [ ]:
'Tokyo' in city_data.axes[0]

In [ ]:
'revenues' in city_data.axes[1]

In [ ]:
'revenue' in city_data.axes[1]

In [ ]:
city_data.keys()

In [ ]:
"Amsterdam" in city_data

In [ ]:
"revenue" in city_data

-----------



Accessing Series Elements

Pada bagian di atas, kalian telah membuat Series Pandas berdasarkan list Python dan membandingkan kedua struktur datanya. Kalian telah melihat kemiripan objek Series dengan list dan dictionaries dalam beberapa cara. Kesamaan lebih lanjut adalah kalian dapat menggunakan operator pengindeksan ([]) untuk Series juga.

Kalian juga akan mempelajari cara menggunakan dua metode akses khusus Pandas:

* .loc
* .iloc


|notes    | loc    | iloc |
|:--------|:-------|:-----| 
|a value   | ```loc['label'] or loc[0]``` <br> single label or integer  |```iloc[1]``` single integer |  
|a list |  ```loc[['label','label2]] ```  <br>  A list of labe | ``` iloc[[1,2,3]]``` list integers  
| slicing |  ```loc[['A':'B']] ```  <br> A and B Include     | ``` iloc[n:m]```, n is include m exclude |
| Condition |   bool series or list        | bool list |
| Callable function | ```loc[lambda A: A[2]] ```  |  ```iloc[lambda A: A[2]] ``` |


Kalian akan melihat bahwa metode akses data ini bisa jauh lebih mudah dibaca daripada operator pengindeksan.

Using the Indexing Operator

Ingatlah bahwa Series memiliki dua indeks:

* A positional or implicit index, which is always a RangeIndex
* A label or explicit index, which can contain any hashable objects

Selanjutnya, kunjungi kembali objek city_revenues:

In [ ]:

city_revenues

Kalian dapat dengan mudah mengakses nilai dalam Series dengan indeks label dan posisi:

In [ ]:
city_revenues["Toronto"]

In [ ]:
city_revenues[1]

Kalian juga dapat menggunakan indeks dan slicing negatif, seperti yang kalian lakukan untuk list:

In [ ]:
city_revenues[-1]

In [ ]:
city_revenues[1:]

In [ ]:
city_revenues["Toronto":]

#### Using .loc and .iloc

Operator pengindeksan ([]) memang mudah digunakan, tetapi ada peringatan. Bagaimana jika labelnya juga angka? Katakanlah kalian harus bekerja dengan objek Series seperti ini:



In [ ]:
colors = pd.Series(
    ["red", "purple", "blue", "green", "yellow"],
    index=[1, 2, 3, 5, 8]
)

In [ ]:
colors

Apakah output dari colors[1]? Untuk indeks posisi,

colors[1] adalah "ungu". Namun, jika kalian menggunakan indeks label, 

maka colors[1] mengacu pada "merah".

Untuk menghindari kebingungan, pustaka Pandas Python menyediakan dua metode akses data:

* .loc refers to the label index.
* .iloc refers to the positional index.

Metode akses data ini jauh lebih mudah dibaca:


In [ ]:
colors[1]  ## mengambil key / label 1 

In [ ]:
colors.loc[1]  ## mengambil key / label 1 

In [ ]:
colors.iloc[1]  ## mengambil indeks 1 

```colors.loc[1]``` mengembalikan "red", elemen dengan label 1. 

```colors.iloc[1]``` mengembalikan "purple", elemen dengan indeks 1.

Lebih mudah mengingat perbedaan antara .loc dan .iloc daripada mencari tahu apa yang akan dikembalikan oleh operator pengindeksan. Meskipun kalian terbiasa dengan semua kebiasaan operator pengindeksan, dapat berbahaya untuk berasumsi bahwa setiap orang yang membaca kode kalian telah menginternalisasi aturan tersebut juga.

```.loc dan .iloc``` juga mendukung fitur yang kalian harapkan dari operator pengindeksan, seperti slicing. 

Namun, metode akses data ini memiliki perbedaan penting. 

Sementara ```.iloc``` mengecualikan elemen penutup, 

sedangkan ```.loc``` menyertakannya. Lihatlah blok kode ini:

In [ ]:
colors

In [ ]:
# Return the elements with the implicit index: 1, 2
colors.iloc[1:3]

Jika kalian membandingkan kode ini dengan contoh di atas, maka Anda dapat melihat bahwa colors.iloc[1:3] mengembalikan elemen dengan indeks posisi 1 dan 2. Item penutup "hijau" dengan indeks posisi 3 dikecualikan.

Di sisi lain, .loc menyertakan elemen penutup:

In [ ]:
colors.loc?

In [ ]:
# Return the elements with the explicit index between 3 and 8
colors.loc[3:8]

Blok kode ini mengatakan untuk mengembalikan semua elemen dengan indeks label antara 3 dan 8. Di sini, item penutup "kuning" memiliki indeks label 8 dan disertakan dalam keluaran.

Kalian juga dapat menyertakan indeks posisi negatif ke .iloc:

In [ ]:
colors.iloc[-2]

Kalian mulai dari akhir Series dan mengembalikan elemen kedua.

Kalian dapat menggunakan blok kode di atas untuk membedakan antara dua perilaku Series:

```.iloc``` pada series sama penggunaan dg list ```[]```

```.loc``` pada series sama penggunaan dg Dictionary  ```{}```


-------

-------

## Accessing DataFrame Elements

Karena DataFrame terdiri dari objek Series, kalian dapat menggunakan alat yang sama untuk mengakses elemennya. 

Perbedaan krusialnya adalah dimensi tambahan dari DataFrame. Kalian akan menggunakan operator pengindeksan untuk kolom dan metode akses .loc and .iloc di baris.

Using the Indexing Operator

Jika kalian menganggap DataFrame sebagai dictionary yang nilainya adalah Series, maka masuk akal jika kalian dapat mengakses kolomnya dengan operator pengindeksan:

In [ ]:
city_data

In [ ]:
city_data["revenue"]

Di sini, kalian menggunakan operator pengindeksan untuk memilih kolom berlabel "revenue".

Jika nama kolom adalah string, maka kalian juga dapat menggunakan akses attribute-style dengan ```dot notation```:

In [ ]:
city_data.revenue

city_data["revenue"] dan city_data.revenue mengembalikan output yang sama.

Ada satu situasi di mana mengakses elemen DataFrame dengan notasi titik mungkin tidak berfungsi atau dapat menimbulkan kesalahan.

Adalah saat nama kolom bertepatan dengan atribut DataFrame atau nama method :

In [ ]:
toys = pd.DataFrame([
    {"name": "ball", "shape": "sphere"},
    {"name": "Rubik's cube", "shape": "cube"},
])

In [ ]:
toys

In [ ]:
toys["shape"]

In [ ]:
toys.shape

Operasi pengindeksan toys["shape"] mengembalikan data yang benar, tetapi toys.shape mengembalikan bentuk DataFrame. 

Kalian sebaiknya hanya menggunakan pengaksesan bergaya atribut dalam sesi interaktif atau untuk operasi read. Kalian tidak boleh menggunakannya untuk kode produksi atau untuk memanipulasi data (seperti menentukan kolom baru).

**Gunakan ```.loc and .iloc```**

Mirip dengan Series, DataFrame juga menyediakan metode akses data ```.loc and .iloc.```

Ingat, 

```.loc``` menggunakan label dan

```.iloc``` indeks posisi:

In [ ]:
city_data.loc["Amsterdam"]

In [ ]:
city_data.loc["Tokyo": "Toronto"]

In [ ]:
city_data.loc["Tokyo": "Toronto",'revenue']

In [ ]:
city_data.iloc[1]

Setiap baris kode memilih baris yang berbeda dari city_data:

city_data.loc["Amsterdam"] selects the row with the label index "Amsterdam".

city_data.loc["Tokyo": "Toronto"] selects the rows with label indices from "Tokyo" to "Toronto". Remember, .loc is inclusive.

city_data.iloc[1] selects the row with the positional index 1, which is "Tokyo".

Kalian  telah menggunakan .loc dan.iloc pada struktur data kecil. Sekarang, saatnya berlatih dengan data yang lebih besar! Gunakan metode akses data untuk menampilkan baris kedua hingga terakhir dari dataset nba.

Baris kedua hingga terakhir adalah baris dengan indeks posisi -2. Kalian dapat menampilkannya dengan .iloc:



In [ ]:
city_revenues

In [ ]:
city_revenues.sum()

In [ ]:
city_revenues.max()


In [ ]:
city_revenues.min()


In [ ]:
city_revenues.mean()

In [ ]:
city_revenues.describe()

-------

In [ ]:
df

In [ ]:
df.iloc[-2]

Untuk DataFrame, metode akses data .loc dan .iloc juga menerima parameter kedua. Sementara parameter pertama memilih baris berdasarkan indeks, parameter kedua memilih kolom. Kalian dapat menggunakan parameter ini bersama-sama untuk memilih subset baris dan kolom dari DataFrame kalian:

In [ ]:
city_data.loc["Amsterdam": "Tokyo", "revenue"]

Perhatikan bahwa kalian memisahkan parameter dengan koma (,). 
* Parameter pertama, "Amsterdam" : "Tokyo,", menyatakan untuk memilih semua baris di antara dua label tersebut. 
* Parameter kedua muncul setelah koma dan mengatakan untuk memilih kolom "revenue".

Saatnya untuk melihat konstruksi yang sama bekerja dengan kumpulan data nba yang lebih besar. Pilih semua game di antara label 5555 dan 5559. Kalian hanya tertarik pada nama tim dan skor, jadi pilih juga elemen tersebut.

In [ ]:
 df.loc[5555:5559, ["fran_id", "opp_fran", "pts", "opp_pts"]]

#### Querying Your Dataset
Kalian telah melihat cara mengakses subsets kumpulan data yang sangat besar berdasarkan indeksnya. Sekarang, kalian akan memilih baris berdasarkan nilai di kolom kumpulan data kalian untuk query data Anda. Misalnya, kalian dapat membuat DataFrame baru yang hanya berisi game yang dimainkan setelah 2010:

In [ ]:
df 

In [ ]:
# kita ingin menampilkan diatas 2010
current_decade = df[df["year_id"] > 2010]

In [ ]:
current_decade

In [ ]:
current_decade.shape

In [ ]:
df["year_id"] > 2010

Kalian masih memiliki seluruh 23 kolom, tetapi DataFrame baru kalian hanya terdiri dari baris-baris yang nilai dalam kolom "year_id" lebih besar dari 2010.

Kalian juga dapat memilih baris di mana field tertentu yang bukan null:

In [ ]:
df.info()

In [ ]:
df["notes"].notnull()

In [ ]:
games_with_notes = df[df["notes"].notnull()]
games_with_notes.shape

Hal ini dapat membantu jika kalian ingin menghindari nilai yang hilang/missing values dalam kolom. Kalian juga dapat menggunakan .notna() untuk mencapai tujuan yang sama.

Kalian bahkan dapat mengakses nilai dari tipe data objek sebagai str dan melakukan metode string padanya:

In [ ]:
ers_teams = df[df["fran_id"].str.endswith("ers")]

Kalian menggunakan.str.endswith() untuk memfilter kumpulan data kalian dan menemukan semua game yang nama tim tuan rumah diakhiri dengan "ers".


In [ ]:
ers_teams.shape

In [ ]:
ers_teams

In [ ]:
df[ ( df['team_id']=='BLB') & (df['pts'] > 100 ) & (df['opp_pts'] > 100) ]


Lakukan pencarian untuk game Baltimore dimana kedua tim mencetak lebih dari 100 poin. Dan kalian harus exclude duplicates:

In [ ]:
df[
    (df["_iscopy"] == 0) &
    (df["pts"] > 100) &
    (df["opp_pts"] > 100) &
    (df["team_id"] == "BLB")
]

Di sini, kalian menggunakan df["_iscopy"] == 0 untuk hanya menyertakan entri yang bukan salinan/duplikat.

Output kalian harusnya berisi lima game.

Coba buat query lain dengan beberapa kriteria. Pada musim semi 1992, kedua tim dari Los Angeles harus memainkan pertandingan kandang di lapangan lain. 

Buat kueri set data kalian untuk menemukan dua game tersebut. Kedua tim memiliki ID yang dimulai dengan "LA".

Kalian dapat menggunakan .str untuk menemukan ID tim yang dimulai dengan "LA", dan kalian dapat berasumsi bahwa game yang tidak biasa seperti itu akan memiliki "notes":

In [ ]:
df[
    (df["_iscopy"] == 0) &
    (df["team_id"].str.startswith("LA")) &
    (df["year_id"] == 1992) &
    (df["notes"].notnull())
]

Ketika kalian mengetahui cara query dataset kalian dengan beberapa kriteria, kalian akan dapat menjawab pertanyaan yang lebih spesifik tentang kumpulan data kalian.

Grouping and Aggregating Your Data
Kalian mungkin juga ingin mempelajari fitur lain dari kumpulan data kalian, seperti jumlah, rata-rata, atau nilai rata-rata dari sekelompok elemen. Untungnya, pustaka Pandas Python menawarkan fungsi grouping dan aggregation untuk membantu kalian menyelesaikan tugas ini.

Series memiliki lebih dari dua puluh metode berbeda untuk menghitung statistik deskriptif. Berikut beberapa contohnya:

-------

In [ ]:
points = df["pts"]
points

In [ ]:
points.sum()

In [ ]:
df['fran_id'].unique()
##len(df['fran_id'].unique()


DataFrame dapat memiliki beberapa kolom, yang memperkenalkan kita terhadap kemungkinan baru untuk agregasi, seperti grouping:

In [ ]:
group_fran = df.groupby('fran_id',sort=False) 

In [ ]:
group_fran

In [ ]:
group_fran['pts'].sum()

In [ ]:
year_results = df[df['fran_id'] == 'Bulls'].groupby(['year_id','game_result'])

In [ ]:
year_results

In [ ]:
year_results['game_id'].count()

Secara default, Pandas mengurutkan keys grup selama memanggil ke .groupby(). Jika kalian tidak ingin mengurutkan, berikan sort=False. Parameter ini dapat menghasilkan peningkatan kinerja.

kalian juga dapat mengelompokkan beberapa kolom:

In [ ]:
df[
    (df["fran_id"] == "Spurs") &
    (df["year_id"] > 2010)
].groupby(["year_id", "game_result"])["game_id"].count()

Kalian dapat mempraktikkan dasar-dasar ini dengan latihan. Coba lihat musim Golden State Warriors 2014-15 (year_id: 2015).

Berapa banyak kemenangan dan kekalahan yang mereka cetak selama musim reguler dan babak playoff?

Pertama, kalian dapat mengelompokkan berdasarkan "is_playoffs", lalu berdasarkan result:

In [ ]:
df[
    (df["fran_id"] == "Warriors") &
    (df["year_id"] == 2015)
].groupby(["is_playoffs", "game_result"])["game_id"].count()

is_playoffs=0 menunjukkan hasil untuk musim reguler, dan is_playoffs=1 menunjukkan hasil untuk playoff.

Manipulating Columns
Kalian perlu mengetahui cara memanipulasi kolom set data kalian dalam cara yang berbeda dari proses analisis data. Kalian dapat add dan drop kolom sebagai bagian dari fase cleaning data awal, atau nanti, berdasarkan insights analisis Anda.

Buat salinan DataFrame asli kalian untuk digunakan:

#### Changing Columns in a DataFrame


In [ ]:
df

In [ ]:
nba = df.copy()

In [ ]:
nba.shape

Kalian dapat menentukan kolom baru berdasarkan yang sudah ada:

In [ ]:
nba['pts'] - nba['opp_pts']

In [ ]:
df.pts - df.opp_pts

In [ ]:
# menambahkan kolom baru diffrenece 
nba["difference"] = df.pts - df.opp_pts

In [ ]:
nba.head()

In [ ]:
nba['difference'].max() ## perbedaan point 

In [ ]:
 df.elo_i - df.opp_elo_i

In [ ]:
nba["elo_difference"] =df.elo_i - df.opp_elo_i

In [ ]:
nba.shape

In [ ]:
nba.head()

Di sini, kalian menggunakan kolom "pts" dan "opp_pts" untuk membuat kolom baru yang disebut "difference". Kolom baru ini memiliki fungsi yang sama dengan yang lama:

In [ ]:
nba["elo_difference"].max()

In [ ]:
nba["elo_difference"].argmax()


In [ ]:
nba["elo_difference"].argmin()
#nba.iloc[nba["elo_difference"].argmax()]

In [ ]:
nba.iloc[nba['difference'].argmax()]

Di sini, kalian menggunakan fungsi agregasi .max() untuk menemukan nilai terbesar kolom baru kalian.

Kalian juga dapat mengganti nama kolom set data kalian. Tampaknya "game_result" dan  "game_location" terlalu panjang, jadi ganti namanya:

In [ ]:
# nba = nba.rename(
#     columns={
#         "game_result": "result", 
#         "game_location": "location"
#     }, 
#     inplace=True
# )

nba.rename(
    columns={
        "game_result": "result", 
        "game_location": "location"
    },
    inplace=True
)

nba.head()

**.rename()**

    secara default, mengembalikan DataFrame yang diperbarui, membiarkan aslinya tidak terubah. 

    Namun, Anda dapat memberikan argumen kata kunci inplace dan menyetelnya ke True.

**inplace=True**

```nba = nba.rename(
     columns={
         "game_result": "result", 
         "game_location": "location"
     }, 
     inplace=True
 )``` 
 

In [ ]:
nba.info()

In [ ]:
nba.head()

Kumpulan data kalian mungkin berisi kolom yang tidak kalian butuhkan. Misalnya, rating Elo mungkin merupakan konsep yang menarik bagi sebagian orang, tetapi kalian tidak akan menganalisisnya dalam sesi ini. Kalian dapat menghapus empat kolom yang terkait dengan Elo:

In [ ]:
nba.shape

In [ ]:
elo_columns = ["elo_i", "elo_n", "opp_elo_i", "opp_elo_n"]

In [ ]:
nba.drop(elo_columns)

In [ ]:
nba.drop?

gambar column 

Sumbu 0 adalah baris dan sumbu 1 adalah kolom. Secara default,

Default : 
* axis 0 baris 
* axis 1 kolom 


sumbunya adalah 0, jadi .drop() akan mencoba mecari baris dengan label di elo_columns, dan itu tidak ada.

In [ ]:
nba.drop(elo_columns, axis=1)

In [ ]:
nba.info() # apakah berubah 

In [ ]:
## inplace akan merubah asli datanya
nba.drop(elo_columns, inplace=True, axis=1)

In [ ]:
nba.info()

In [ ]:
nba.shape

### Specifying Data Types

Saat kalian membuat DataFrame baru, baik dengan memanggil konstruktor atau membaca file CSV, Pandas menetapkan tipe data ke setiap kolom berdasarkan nilainya. Meskipun pandas melakukan pekerjaan yang cukup baik, namun tidak sempurna. Jika kalian memilih jenis data yang tepat untuk kolom kalian di awal, kalian dapat meningkatkan kinerja kode kalian secara signifikan.

Lihat lagi kolom dari dataset nba:



In [ ]:
nba.info()

Sepuluh kolom kalian memiliki objek tipe data. Sebagian besar kolom objek ini berisi teks arbitrer, tetapi ada juga beberapa kandidat untuk dilakukan konversi tipe data. Misalnya, lihat kolom date_game:

In [ ]:
nba["date_game"] = pd.to_datetime(nba["date_game"])
nba.info()

Di sini, kalian menggunakan .to_datetime() untuk menetapkan semua tanggal game sebagai objek datetime.

Kolom lain berisi teks yang sedikit lebih terstruktur. Kolom game_location hanya dapat memiliki tiga nilai berbeda:

In [ ]:
#nba["game_location"].nunique()
nba["location"].unique()

Tipe data apa yang akan kalian gunakan dalam database relasional untuk kolom seperti itu? Kalian mungkin tidak akan menggunakan tipe varchar, melainkan enum. Pandas menyediakan tipe data categorical untuk tujuan yang sama:

In [ ]:
nba["location"].value_counts()

In [ ]:
nba["location"] = pd.Categorical(nba["location"])
nba.info()

*Perhatikan pengunaan memory menjadi kecil* 

Perhatikan pada location dtype berubah menjadi category 

In [ ]:
nba["location"].dtype

merubah game result untuk menjadi categorie 

In [ ]:
nba['result'].unique()

In [ ]:
nba["result"] = pd.Categorical(nba["result"])
nba.info()

Data kategorikal memiliki beberapa keunggulan dibandingkan teks tidak terstruktur. Saat kalian menentukan tipe data kategorikal, kalian membuat validasi lebih mudah dan menghemat banyak memori, karena Pandas hanya akan menggunakan nilai unik secara internal. Semakin tinggi rasio nilai total terhadap nilai unik, semakin banyak penghematan ruang yang akan kalian dapatkan.

Jalankan df.info() lagi. kalian harusnya melihat bahwa mengubah tipe data game_location dari object ke categorical telah menurunkan penggunaan memori.


-----------


### Missing Values
Pernahkah kalian  bertanya-tanya mengapa .info() menunjukkan berapa banyak nilai bukan null yang berisi kolom? Alasan mengapa ini adalah informasi penting. Null values sering kali menunjukkan masalah dalam proses pengumpulan data. Mereka dapat membuat beberapa teknik analisis, seperti berbagai jenis machine learning, menjadi sulit atau bahkan tidak mungkin.

Saat kalian memeriksa kumpulan data nba dengan nba.info(), kalian akan melihat bahwa itu cukup rapi. Hanya catatan kolom yang berisi nilai null untuk sebagian besar barisnya:

In [ ]:
nba.info()

Output ini menunjukkan bahwa kolom notes hanya memiliki **5424 nilai non-null**. Itu berarti lebih dari 120.000 baris kumpulan data kalian memiliki nilai null di kolom ini.

Terkadang, cara termudah untuk menangani dataset yang berisi missing values adalah dengan mengabaikannya. Kalian dapat menghapus semua baris dengan nilai yang hilang menggunakan .dropna():

In [ ]:
nba.shape

In [ ]:
rows_without_missing_data = nba.dropna()

In [ ]:
rows_without_missing_data.shape

Tentu saja, jenis pembersihan data ini tidak masuk akal untuk kumpulan data nba kalian, karena tidak masalah jika game tidak mempunyai notes. 

Tetapi jika kumpulan data kalian berisi satu juta catatan yang valid dan seratus di mana data yang relevan hilang, maka menghapus catatan yang tidak lengkap bisa menjadi solusi yang masuk akal.

Kalian juga dapat menghapus kolom bermasalah jika tidak relevan untuk analisis kalian. Untuk melakukan ini, gunakan .dropna() lagi dan berikan parameter axis=1:

In [ ]:
data_without_missing_columns = nba.dropna(axis=1)

In [ ]:
data_without_missing_columns.shape

In [ ]:
data_without_missing_columns.info()

Sekarang, DataFrame yang dihasilkan berisi semua 126.314 game.

Jika ada nilai default yang berarti untuk use case kalian, kalian juga dapat mengganti nilai yang hilang dengan:

In [ ]:
data_with_default_notes = nba.copy()

In [ ]:
data_with_default_notes["notes"].fillna(
    value="no notes at all",
    inplace=True
)

In [ ]:
data_with_default_notes["notes"].describe()

Di sini, kalian mengisi baris catatan kosong dengan string "no notes at all".

In [ ]:
data_with_default_notes.info()

perhatikan *notes           126314 non-null* 



#### Invalid Values
Invalid values bisa lebih berbahaya daripada missing values. Seringkali, kalian tetap dapat melakukan analisis data seperti yang diharapkan, tetapi hasil yang kalian peroleh aneh karena invalid values. Ini sangat penting jika kumpulan data kalian sangat besar atau menggunakan entri manual. Nilai yang tidak valid seringkali lebih sulit untuk dideteksi, tetapi kalian dapat menerapkan beberapa pemeriksaan dengan kueri dan agregasi.

Satu hal yang dapat kalian lakukan adalah memvalidasi rentang data kalian. Untuk ini, .describe() cukup berguna. Ingatlah bahwa ia mengembalikan keluaran berikut:

In [ ]:
nba.describe()

year_id bervariasi antara 1947 dan 2015. Kedengarannya masuk akal.

Bagaimana dengan pts? Bagaimana bisa minimumnya menjadi **0** ? Mari kita lihat game-game tersebut:

In [ ]:
nba[nba["pts"] == 0]

*Pertandingan "pelanggaran" kontroversial antara Virginia Squires dan Denver Rockets pada 26 Oktober telah dibatalkan untuk Squires. Robert Carlson, komisaris American Basketball Association, membuat keputusan kemarin.

The Squires memenangkan pertandingan, 155-111, dengan bantuan 74 lemparan bebas. Pelatih Denver Alex Hannum telah menginstruksikan para pemainnya untuk melanggar setiap pemain Virginia yang mencoba melakukan tembakan. Denver melakukan 56 pelanggaran dan tujuh Rockets dilanggar.

Hannum mengatakan dia melakukan percobaan karena **"the trend of pressure defense".**
https://www.nytimes.com/1972/11/07/archives/rocket-foul-game-awarded-to-squires.html

Sepertinya game itu forfeited. Bergantung pada analisis kalian, kalian mungkin ingin menghapusnya daridataset.

Inconsistent Values

Kadang-kadang sebuah nilai akan sepenuhnya realistis, tetapi tidak sesuai dengan nilai di kolom lain.

Dalam kumpulan data NBA, nilai kolom pts, opp_pts dan game_result harus konsisten satu sama lain. kalian dapat memeriksanya menggunakan atribut .empty:

In [ ]:
# pengecekan data Lose tapi pointnya lebih besar dari lawan  
nba[(nba["pts"] > nba["opp_pts"]) & (nba["result"] != 'W')]

In [ ]:
nba[(nba["pts"] > nba["opp_pts"]) & (nba["result"] != 'W')].empty

Jika hasil pertandingan adalah kemenangan ``` game_result = W ```, 

maka poin ```pts``` yang dicetak untuk baris tersebut harus lebih banyak ```> ``` dari poin lawan ```opp_pts``` yang dicetak. 

Dan untuk kekalahan, poin lawan harus lebih besar.

In [ ]:
nba[(nba["pts"] < nba["opp_pts"]) & (nba["result"] != 'L')].empty

Untungnya, kedua kueri ini mengembalikan DataFrame kosong.

Bersiaplah untuk kejutan setiap kali kali bekerja dengan kumpulan data mentah, terutama jika kumpulan data tersebut dikumpulkan dari sumber berbeda atau melalui pipeline yang rumit. Kalian mungkin melihat baris di mana team mencetak lebih banyak poin daripada lawannya, tetapi masih kalah.

------

### Combining Multiple Datasets
Di bagian sebelumnya, kalian telah mempelajari cara membersihkan kumpulan data yang berantakan. Aspek lain dari data dunia nyata adalah sering kali data datang dalam beberapa bagian. Di bagian ini, kalian akan mempelajari cara mengambil bagian-bagian tersebut dan menggabungkannya menjadi satu kumpulan data yang siap untuk dianalisis.

Sebelumnya, kalian menggabungkan dua objek Series ke dalam DataFrame berdasarkan indeksnya. 

Sekarang, kalian akan mengambil satu langkah lebih jauh dan menggunakan **.concat()** untuk menggabungkan city_data dengan DataFrame lain. 

Misalkan kalian berhasil mengumpulkan beberapa data di dua kota lagi:

In [ ]:
city_data

In [ ]:
further_city_data = pd.DataFrame(
    {
        "revenue": [7000, 3400], 
        "employee_count": [2, 2]
    },
    index=["New York", "Barcelona"]
)

In [ ]:
further_city_data

DataFrame kedua ini berisi info tentang kota "New York" dan "Barcelona".

Kalian dapat menambahkan kota-kota ini ke city_data menggunakan .concat():

In [ ]:
all_city_data = pd.concat([city_data, further_city_data], sort=False)

In [ ]:
all_city_data

Sekarang, variabel baru all_city_data  berisi nilai dari kedua objek DataFrame.

Secara default, concat()  menggabungkan sepanjang **axis = 0**. Dengan kata lain, concat menambahkan baris. Kalian juga dapat menggunakannya untuk menambahkan **kolom** dengan menyediakan parameter **axis = 1**:

In [ ]:
city_countries = pd.DataFrame({
    "country": ["Holland", "Japan", "Holland", "Canada", "Spain"],
    "capital": [1, 1, 0, 0, 0]},
    index=["Amsterdam", "Tokyo", "Rotterdam", "Toronto", "Barcelona"]
)


In [ ]:
city_countries

In [ ]:
all_city_data

In [ ]:
pd.concat?

In [ ]:
cities = pd.concat([all_city_data, city_countries], axis=1, sort=False)

In [ ]:
cities

Perhatikan bagaimana Pandas menambahkan NaN untuk nilai yang hilang.

Jika kalian ingin menggabungkan hanya kota yang muncul di kedua objek DataFrame, maka kalian dapat mengatur parameter join menjadi inner:

In [ ]:
pd.concat([all_city_data, city_countries], axis=1, join="inner")

Toronto	8000.0	NaN	Canada	0.0
Rotterdam	NaN	NaN	Holland	0.0


In [ ]:
pd.concat([all_city_data, city_countries], axis=1, join="outer")

Seperti yang Anda lihat, DataFrame gabungan berisi baris untuk 'New York' dan 'Barcelona'. Pastikan untuk secara eksplisit mengatur argumen kata kunci sortir. Itu tidak diperlukan, tetapi Pandas baru-baru ini mengubah nilai default dari True menjadi False. Sampai versi baru digunakan secara luas, menyetel argumen kata kunci secara eksplisit akan membantu menghindari kebingungan.

Perhatikan bahwa fungsi concat() menggabungkan DataFrames menggunakan baris. Dengan menyetel argumen kata kunci sumbu ke 1, Anda dapat menggabungkan kolom.

Misalkan Anda memiliki DataFrame baru dengan kolom berbeda tetapi indeksnya sama dengan all_city_data DataFrame.

Sekarang Anda dapat memanggil concat(), berikan daftar DataFrames untuk digabungkan, dan atur sumbu ke 1 untuk menambahkan kolom baru ke DataFrame.

Perhatikan NaN yang mewakili nilai yang hilang di DataFrame. Untuk menghilangkannya, setel argumen kata kunci join ke 'inner'. Gabung dalam hanya akan menyimpan baris dengan indeks di kedua DataFrames.

Ini tidak menghapus NaN untuk 'Toronto', karena indeks untuk 'Toronto' masih ada di kedua DataFrames. Secara default, gabungan adalah gabungan luar, yang mencakup semua baris.

-------

Meskipun paling mudah untuk menggabungkan data berdasarkan indeks, itu bukan satu-satunya kemungkinan. Anda dapat menggunakan .merge() untuk mengimplementasikan operasi gabungan yang mirip dengan operasi dari SQL:

Meskipun paling mudah untuk menggabungkan data berdasarkan indeks, itu bukanlah satu-satunya cara. Kalian bisa menggunakan .merge() untuk mengimplementasikan operasi gabungan yang mirip dengan SQL:While it’s most straightforward to combine data based on the index, it’s not the only possibility. You can use .merge() to implement a join operation similar to the one from SQL:

Meskipun paling mudah untuk menggabungkan data berdasarkan indeks, itu bukanlah satu-satunya cara. Kalian bisa menggunakan .merge() untuk mengimplementasikan operasi gabungan yang mirip dengan SQL:

In [ ]:
countries = pd.DataFrame({
    "population_millions": [17, 127, 37],
    "continent": ["Europe", "Asia", "North America"]
}, index=["Holland", "Japan", "Canada"])


Di sini, kalian meneruskan parameter left_on="country" ke .merge() untuk menunjukkan kolom mana yang ingin kalian join. Hasilnya adalah DataFrame yang lebih besar yang tidak hanya berisi data kota, tetapi juga populasi dan benua dari masing-masing negara:

In [ ]:
countries

In [ ]:
cities

In [ ]:
pd.merge(cities, countries, left_on="country", right_index=True)

Perhatikan bahwa hasilnya hanya berisi kota-kota di mana negara tersebut dikenal dan muncul di DataFrame yang digabungkan.

Newyork 
Barcelona 

.merge() melakukan inner join secara default. Jika kalian ingin memasukkan semua kota dalam hasil, maka kalian perlu memberikan parameter how:

In [ ]:
pd.merge(
    cities,
    countries,
    left_on="country",
    right_index=True,
    how="left"
)

Dengan gabungan kiri ini, Anda akan melihat semua kota, termasuk yang tidak memiliki data negara.

Dengan left join ini, kalian akan melihat semua kota, termasuk kota yang tidak memiliki data negara.

Visualizing Your Pandas DataFrame
Visualisasi data adalah salah satu hal yang bekerja jauh lebih baik di notebook Jupyter daripada di terminal.

Sertakan baris ini untuk menampilkan plot langsung di notebook:

In [ ]:
pip install matplotlib

In [ ]:
%matplotlib inline

Objek Series dan DataFrame memiliki metode .plot(), yang merupakan wrapper pada matplotlib.pyplot.plot(). Secara default, syntax tersebut membuat line plot. Visualisasikan berapa banyak poin yang dicetak Knicks sepanjang musim:

In [ ]:
nba.head()

In [ ]:
knick_pts_by_year = nba[nba["fran_id"] == "Knicks"].groupby("year_id")["pts"].sum()

In [ ]:
knick_pts_by_year

In [ ]:
knick_pts_by_year.plot()

In [ ]:
knick_pts_by_year.plot(kind='bar')

In [ ]:
knick_pts_by_year.tail(10).plot(kind='bar')

In [ ]:
knick_year_2000 = nba[ ( nba["fran_id"] == "Knicks") & (nba['year_id']>2000)]

#.plot(kind='scater',x='pts')
knick_year_2000

In [ ]:
#
knick_year_2000.plot(kind='scatter',x='pts',y='opp_pts')

In [ ]:
knick_year_2000.plot(kind='scatter',x='pts',y='opp_pts',s=0.5)

In [ ]:
nba[nba['pts'] == nba['opp_pts']].empty

In [ ]:
heat_13_w1 = nba[
    (nba['fran_id']=='Heat') & 
    (nba['year_id'] ==2013)
]['result'].value_counts()

heat_13_w1

In [ ]:
heat_13_w1.plot(kind='pie')

In [ ]:
nba[nba["fran_id"] == "Knicks"].groupby("year_id")["pts"].sum()

In [ ]:
nba["fran_id"].value_counts().head(10).plot(kind="bar")

In [ ]:
nba["fran_id"].value_counts().head(10)

#### Summary 

* Pandas is python package for Exploratory Data Analis 
* The Series and DataFrame 
* Access specific values with .loc , .iloc and indexing 
* Summarize data with grouping anda aggregation methods
* Write queries to get precise result 
* Manage missing values 
* Visualizations 

## Latihan Menggunakan Data Set Iris 

https://raw.githubusercontent.com/jbrownlee/Datasets/master/iris.csv
#### Load dataset 
#### Tampilkan 5 baris pertama 
#### Tampilkan 4 baris pertama 
#### Describe dataframe 
#### Hapus colum id 
#### berapa baris dan kolom data set iris ? 
#### Mengambil list dari semua kolom 
#### Mengambil Basic Information
#### Identifikasi Nilai yang missing 
#### Identifikasi yang duplikat baris / kolom
#### menghitung jumlah duplikat baris / kolom
#### Menghapus duplikasi 

In [1]:
pip install dtale

     |████████████████████████████████| 12.8 MB 782 kB/s            
  Using cached dash_colorscales-0.0.4-py3-none-any.whl
  Using cached kaleido-0.2.1-py2.py3-none-macosx_10_11_x86_64.whl (85.2 MB)
     |████████████████████████████████| 216 kB 927 kB/s            
  Using cached missingno-0.4.2-py3-none-any.whl (9.7 kB)
  Using cached future-0.18.2-py3-none-any.whl
  Using cached strsimpy-0.2.1-py3-none-any.whl (45 kB)
  Using cached flask_ngrok-0.0.25-py3-none-any.whl (3.1 kB)
     |████████████████████████████████| 9.9 MB 751 kB/s            
     |████████████████████████████████| 927 kB 554 kB/s            
  Using cached dash_daq-0.5.0-py3-none-any.whl
     |████████████████████████████████| 90 kB 634 kB/s            
  Using cached squarify-0.4.3-py3-none-any.whl (4.3 kB)
     |████████████████████████████████| 2.0 MB 496 kB/s            
  Using cached dash_table-5.0.0-py3-none-any.whl (3.9 kB)
  Using cached dash_html_components-2.0.0-py3-none-any.whl (4.1 kB)
  Using cached

In [ ]:
import dtale 

In [2]:
data = pd.read_csv("data/kapal_titanic.csv")

NameError: name 'pd' is not defined

In [ ]:
data.head()

In [ ]:
data.tail()

In [ ]:
data.to_csv("mydata.csv",index=False)

In [ ]:
ls ../data/

In [ ]:
ls 

In [ ]:
mydata = pd.read_csv("mydata.csv")

In [ ]:
mydata.head()

In [ ]:
mydata.to_csv("mydataNew.csv",index=True)

In [ ]:
myDataNew = pd.read_csv("mydataNew.csv")
myDataNew.head() ## walaupun tdk terlihat index coba di cek di file

----------------------------

# Pandas & Database

## Preparation Setting Database 

### Create database mysql 
    *   mysql> create database h8pytn071;

### Create User database 
    *  mysql> CREATE USER 'userpython'@'%' IDENTIFIED BY 'H8pytn071';

### Grant Prefileges user to database 
    *  mysql> GRANT ALL PRIVILEGES ON h8pytn071.*  TO 'userH8pytn071'@'%';
### flush 
    *  FLUSH PRIVILEGES;


In [ ]:
pip install sqlalchemy

In [ ]:
pip install pymysql

In [ ]:
pip install pandas

In [ ]:
pip install cryptography


In [ ]:
from sqlalchemy import create_engine

import pymysql

import pandas as pd



In [ ]:
tableName2   = "usertrans3"

In [ ]:

 

data_user_trans = {"UserId":["xxxxx", "yyyyy", "zzzzz", "aaaaa", "bbbbb", "ccccc", "ddddd"],

            "UserFavourite":["Greek Salad", "Philly Cheese Steak", "Turkey Burger", "Crispy Orange Chicken", "Atlantic Salmon", "Pot roast", "Banana split"],

            "MonthlyOrderFrequency":[5, 1, 2, 2, 7, 6, 1],

            "HighestOrderAmount":[30, 20, 16, 23, 20, 26, 9],

            "LastOrderAmount":[21,20,4,11,7,7,7],

            "LastOrderRating":[3,3,3,2,3,2,4],

            "AverageOrderRating":[3,4,2,1,3,4,3],

            "OrderMode":["Web", "App", "App", "App", "Web", "Web", "App"],

            "InMedicalCare":["No", "No", "No", "No", "Yes", "No", "No"]};

In [ ]:
dataFrame   = pd.DataFrame(data=data_user_trans)  

In [ ]:
dataFrame.head()

In [ ]:
sqlEngine       = create_engine('mysql+pymysql://userH8pytn071:PwdH8pytn071@localhost:6603/h8pytn071', pool_recycle=3600)

In [ ]:
dbConnection    = sqlEngine.connect()



In [ ]:
try:

    frame  = dataFrame.to_sql(tableName2, dbConnection, if_exists='fail');

except ValueError as vx:

    print(vx)

except Exception as ex:   

    print(ex)

else:

    print("Table %s created successfully."%tableName2);   

finally:

    dbConnection.close()

### periksa tabel pada database di mysql 
    * mysql> use h8pytn071
    * mysql> show tables;
    * mysql> desc usertrans;

In [ ]:
 from sqlalchemy import create_engine

import pymysql

import pandas as pd

 

sqlEngine       = create_engine('mysql+pymysql://userH8pytn071:PwdH8pytn071@localhost:6603/h8pytn071', pool_recycle=3600)

dbConnection    = sqlEngine.connect()

frame           = pd.read_sql("select * from h8pytn071.usertrans3", dbConnection);

 

pd.set_option('display.expand_frame_repr', False)

print(frame)

 

dbConnection.close()

-----------

## Pandas from csv to database 

In [ ]:
from sqlalchemy import create_engine

import pymysql

import pandas as pd

In [ ]:
data = pd.read_csv('data/Sample-Superstore.csv')


In [ ]:
data.head()

In [ ]:
sqlEngine       = create_engine('mysql+pymysql://userH8pytn071:PwdH8pytn071@localhost:6603/h8pytn071', pool_recycle=3600)

In [ ]:
dbConnection    = sqlEngine.connect()

In [ ]:
tableName ="superstore2"

In [ ]:
frame  = dataFrame.to_sql(tableName, dbConnection, if_exists='replace');


## Read File Xls 


In [ ]:
pip install openpyxl

In [ ]:
df_can = pd.read_excel('https://github.com/ardhiraka/PFDS_sources/blob/master/Canada.xlsx?raw=true',
                       sheet_name='Canada by Citizenship',
                       skiprows=range(20),
                       skipfooter=2) #engine='openpyxl'

                  

print ('Data read into a pandas dataframe!')

In [ ]:
df_can.head

In [ ]:
dataFrame = pd.DataFrame({'Name': [' RACHEL  ', ' MONICA  ', ' PHOEBE  ',
                                   '  ROSS    ', 'CHANDLER', ' JOEY    '],
                            
                          'Age': [30, 35, 37, 33, 34, 30],
                            
                          'Salary': [100000, 93000, 88000, 120000, 94000, 95000],
                            
                          'JOB': ['DESIGNER', 'CHEF', 'MASUS', 'PALENTOLOGY',
                                  'IT', 'ARTIST']})

In [ ]:
# filter dataframe 
display(dataFrame[dataFrame.eval("Salary <=100000 & (Age <40) & JOB.str.startswith('A').values")])

In [ ]:
df = pd.read_csv('IMDB.csv', encoding = "ISO-8859-1", skip_blank_lines=False)

In [ ]:
df = pd.read_csv('IMDB.csv', encoding = "ISO-8859-1", skiprows = [1,3,7])

In [ ]:
df.tail(2)
df = pd.read_csv('IMDB.csv', encoding = "ISO-8859-1", skipfooter=2, engine='python')
df.tail(2)

In [ ]:
df = pd.read_csv('IMDB.csv', encoding = "ISO-8859-1", nrows=100) 
df.shape

In [ ]:
df = pd.read_excel('IMDB.xlsx', sheetname= 0, na_values=[' '])

In [3]:
import pandas as pd 

df = pd.read_csv("penjualan.csv")

In [4]:
df

Unnamed: 0  Batagor  Simomay
0       Budi        3        0
1       Wati        2        5
2       Iwan        0        6
3       Joko        2        1

In [5]:
df.sort_index(axis=1)

Batagor  Simomay Unnamed: 0
0        3        0       Budi
1        2        5       Wati
2        0        6       Iwan
3        2        1       Joko

In [15]:
df.sort_values(by=['Simomay' , 'Batagor'])



Unnamed: 0  Batagor  Simomay
0       Budi        3        0
3       Joko        2        1
1       Wati        2        5
2       Iwan        0        6

In [20]:
df.sort_values(by=['Batagor'],axis=0)

Unnamed: 0  Batagor  Simomay
2       Iwan        0        6
1       Wati        2        5
3       Joko        2        1
0       Budi        3        0

In [21]:
df.sort_index(0)

/var/folders/gh/bm95p3ys6sd6jvzbfl061_3c0000gn/T/ipykernel_43808/2388537916.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.sort_index will be keyword-only
  df.sort_index(0)


Unnamed: 0  Batagor  Simomay
0       Budi        3        0
1       Wati        2        5
2       Iwan        0        6
3       Joko        2        1